osu.Game/Scoring/ScoreManager.cs

```diff
@@ -43,6 +43,8 @@ public class ScoreManager : DownloadableArchiveModelManager<ScoreInfo, ScoreFile
         [CanBeNull]
         private readonly OsuConfigManager configManager;
 
+        public Action<LegacyScoreDecoder.BeatmapNotFoundException, ArchiveReader> OnMissingBeatmap;
```

Comment:
This event signature is all sorts of wrong. Read whatever info you need *inside* the score manager and only *then* push *that* out, instead of exposing internal exception types and archives outside to the base game.

Also, passing data (beatmap hashes & score info) around via instantiated exceptions is pretty bad in my opinion.

Reply:
So `onMissingBeatmap(string beatmapHash, string message, Score score, ArchiveReader archive)`?

Exposing archive is for importing the replay again after downloading the beatmap from the replay import screen. Or should I expose the path?